### ISSUE TITLE
Open files on a mounted share receiving updates only every 90 seconds

### ISSUE THREAD

**Issue by trantor** [nfs-ganesha/nfs-ganesha#294]
Hello.

I've exported through ganesha a folder on an ext4 fs.
It's an NFS4 share which is used to access log files from a second machine.
Both machines are running Ubuntu 16.04 and the version of ganesha running is 2.6.0 .

The log files which the client machine needs to access are updated multiple times per second, but I've noticed, tailing them from the client machine, that the new lines arrive in chunks roughly every 90 seconds.
On top of that I've noticed how the nfs daemon leaves open file descriptors to logs which have been rotated out of existence weeks in the past.

I am guessing that this is related to caching somewhat and I've tried to tweak related parameters in the ganesha configuration file, without much success.
Can anyone shed some light on these matters?

Thanks in advance.

**Comment by dang** [comment]
Ganesha caches attributes, to avoid spamming filesystems with getattr calls (which happen in almost every OP in NFSv4).  The default cache time is 60 seconds, and VFS has no way of notifying the cache of a change in the file, so attributes like change time and size won't update on the clients for at least 60 seconds.  This is configurable via the Attr_Expiration_Time config entry in the Export (or Export_Defaults for the entire box).  It can be set to 0 to disable attribute caching entirely.

FDs for NFSv4 clients are kept open as long as the client keeps them open.  In NFSv3, there weren't "open" or "close" operations, so FDs are opened on each read/write/getattr/setattr and closed over time.  But for NFSv4, there are explicit Open and Close OPs, so the FDs are left open as long as the client keeps them open.

**Comment by trantor** [comment]
Hello again.
Below my new export configuration with Attr_Expiration_Time set to 0 .

```
NFS_CORE_PARAM {
        mount_path_pseudo = true;
        Protocols = 4;
        Sectype = sys;
}

EXPORT
{
        Export_Id = 12345;
        Path = /opt/zimbra/log;
        Pseudo = /log;
        Access_Type = None;
        Attr_Expiration_Time = 0;
        Squash = None;

        Sectype = sys;
        Transports = TCP;

        FSAL {
               Name = VFS;
        }

        CLIENT {
                Clients = 10.69.95.178;
                Access_Type = RO;
        }
}
```
What I see is as follows.
After mounting the share the first access (e.g. ls /mnt/share_mountpoint) takes a few dozen seconds to output anything.
After that data starts to flow, so to speak.
This time the behaviour is similar to what described at the beginning, but the wait between one chunk of data and the next has gone down to aroun 3 seconds.

In other words, doing a `tail -f /mnt/share_mountpoint/logfile` will get me a bunch of lines, then nothing for 3 seconds, then a new bunch of lines, then nothing for 3 seconds, then a new bunch of lines and so on.

Is that supposed to happen with Attr_Expiration_Time set to 0? BTW before, during the 90 seconds of wait between updates I think that modification times actually were updated and seen as such on the client's side.

About the open FDs I don't think that the process accessing them through the share kept them open after the logs were rotated, but I'll verify that empirically.

**Comment by ffilz** [comment]
Closing this due to age, feel free to re-open if it's still of interest and you have explored using a currently supported version of Ganesha (V2.8 or V3 at this time).
